lib/ts/nextjs/middleware.ts

```diff
@@ -0,0 +1,204 @@
+    }
+
+    const requestUrl = new URL(request.url);
+    const redirectTo = requestUrl.searchParams.get(REDIRECT_PATH_PARAM_NAME) || "/";
+    try {
+        const tokens = await fetchNewTokens(refreshToken);
+        const hasRequiredCookies = tokens.accessToken.cookie && tokens.refreshToken.cookie && tokens.frontToken.cookie;
+        const hasRequiredHeaders = tokens.accessToken.header && tokens.refreshToken.header && tokens.frontToken.header;
+        if (!hasRequiredCookies && !hasRequiredHeaders) {
+            logDebugMessage("Missing tokens from refresh response");
+            return redirectToAuthPage(request);
+        }
+        const redirectUrl = new URL(redirectTo, request.url);
+        const finalResponse = new Response(null, {
+            status: 307,
+            headers: {
+                Location: redirectUrl.toString(),
+            },
+        });
+        finalResponse.headers.append(FRONT_TOKEN_HEADER_NAME, tokens.frontToken.header as string);
+        if (hasRequiredCookies) {
+            finalResponse.headers.append("set-cookie", `${tokens.accessToken.cookie as string}`);
+            finalResponse.headers.append("set-cookie", `${tokens.refreshToken.cookie as string}`);
+            finalResponse.headers.append("set-cookie", `${tokens.frontToken.cookie as string}`);
+        }
+        if (hasRequiredHeaders) {
+            finalResponse.headers.append(REFRESH_TOKEN_HEADER_NAME, tokens.refreshToken.header as string);
+            finalResponse.headers.append(ACCESS_TOKEN_HEADER_NAME, tokens.accessToken.header as string);
+        }
+
+        if (tokens.antiCsrfToken.cookie) {
+            finalResponse.headers.append("set-cookie", `${tokens.antiCsrfToken.cookie as string}`);
+        } else if (tokens.antiCsrfToken.header) {
+            finalResponse.headers.append(ANTI_CSRF_TOKEN_HEADER_NAME, tokens.antiCsrfToken.header);
+        }
+        finalResponse.headers.append(
+            "set-cookie",
+            `${REDIRECT_ATTEMPT_COUNT_COOKIE_NAME}=${
+                redirectAttemptNumber + 1
+            }; Path=/; Max-Age=10; HttpOnly; SameSite=Strict`
+        );
+        logDebugMessage("Attached new tokens to response");
+        return finalResponse;
+    } catch (err) {
+        logDebugMessage("Error refreshing session");
+        logDebugMessage(err as unknown as string);
+        return redirectToAuthPage(request);
+    }
+}
+
+export async function revokeSession(config: SuperTokensNextjsConfig): Promise<Response | void> {
+    AppInfo = config.appInfo;
+    if (config.enableDebugLogs) {
+        enableLogging();
+    }
+    const signOutURL = new URL(`${AppInfo.apiBasePath}/signout`, AppInfo.apiDomain);
+    try {
+        const signOutResponse = await fetch(signOutURL, {
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+            },
+            credentials: "include",
+        });
+
+        if (signOutResponse.status !== 200) {
```

Comment:
This can be a 401, which would require a retry

Reply:
resolved